### PR TITLE
Adding drop down to diff table

### PR DESCRIPTION
**Pull request by carolynqu**
For deferred apps, in the diff-table, the tab auto-opens to "Entire app" with a dropdown to switch between the entire app, the main unit, and the deferred units:

https://user-images.githubusercontent.com/67798872/183158040-548f1261-11b0-4201-aec1-811b114ca01a.mp4

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart [hunk 1]
```diff
@@ -458,25 +519,91 @@ class AppSizeController {
     diffMap['n'] = isDeferredApp.value ? _entireAppNodeName : _rootNodeName;
 
     // TODO(peterdjlee): Try to move the non-active tree generation to separate isolates.
-    _combinedDiffTreeRoot = generateDiffTree(
-      diffMap,
-      DiffTreeType.combined,
-      skipNodesWithNoByteSizeChange: !isDeferredApp.value,
+    // Entire app or root (for non-deferred):
+    final diffTreeMap = _generateDiffTrees(diffMap);
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart:523]: Let's actually set the `diffTreeMap` (and the `mainDiffTreeMap`, etc) as a top-level variable. So instead of:
```
    _combinedDiffTreeRoot
    _increasedDiffTreeRoot
    _decreasedDiffTreeRoot
  _mainCombinedDiffTreeRoot
   etc..
``` 
We only have:

```
_diffTreeMap
_mainDiffTreeMap
_deferredDiffTreeMap
```

File: packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart [hunk 2]
```diff
@@ -161,21 +175,64 @@ class AppSizeController {
...
   TreemapNode? get _activeDiffRoot {
-    switch (_activeDiffTreeType.value) {
-      case DiffTreeType.increaseOnly:
-        return _increasedDiffTreeRoot;
-      case DiffTreeType.decreaseOnly:
-        return _decreasedDiffTreeRoot;
-      case DiffTreeType.combined:
-      default:
-        return _combinedDiffTreeRoot;
+    final isDeferred = _isDeferredApp.value;
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart:178]: If we follow the suggestion I made below, we can simplify this getter considerably. We can add another getter for the `_activeDiffMap` which returns either the `_diffMap`, `_mainDiffMap`, or `_deferredDiffMap`. Then, this `getter` only needs to return either `_activeDiffMap.combined`, `_activeDiffMap.increaseOnly` or `_activeDiffMap.decreaseOnly`

Comment (elliette) [reply packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart]: Looks better, but adding an `_activeDiffMap` getter (as described above) will clean this up even more!

File: packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart [hunk 3]
```diff
@@ -160,21 +174,39 @@ class AppSizeController {
...
+      case AppUnit.entireApp:
+      default:
+        return _diffTreeMap;
+    }
+  }
+
   TreemapNode? get _activeDiffRoot {
-    switch (_activeDiffTreeType.value) {
+    final diffTreeType = _activeDiffTreeType.value;
+    final diffMap = _activeDiffMap;
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart:193]: Let's return `null` if `_activeDiffMap` is `null` (instead of null-asserting with `!`)

**Comment by elliette** [comment]
Let's add a test case for the diff view with deferred units (it should be pretty similar to the test here: https://github.com/flutter/devtools/pull/4332)

File: packages/devtools_app/test/app_size/app_size_screen_test.dart [hunk 4]
```diff
@@ -53,6 +55,18 @@ void main() {
...
   );
 
+  final diffDeferredAppFile = DevToolsJsonFile(
+    name: 'lib/src/app_size/stub_data/diff_deferred_app.dart',
+    lastModifiedTime: lastModifiedTime,
+    data: json.decode(diffDeferredApp),
+  );
+
+  final diffNonDeferredAppFile = DevToolsJsonFile(
+    name: 'lib/src/app_size/stub_data/diff_deferred_app.dart',
```
Comment (elliette) [inline packages/devtools_app/test/app_size/app_size_screen_test.dart:65]: Should this be `diff_no_deferred_app`?

**Review by elliette — APPROVED** [review]

File: packages/devtools_app/lib/src/screens/app_size/app_size_screen.dart [hunk 5]
```diff
@@ -208,9 +208,7 @@ class _AppSizeBodyState extends State<AppSizeBody>
                   ),
                   Row(
                     children: [
-                      if (currentTab.key == AppSizeScreen.analysisTabKey &&
-                          isDeferredApp)
-                        _buildAppUnitDropdown(),
+                      if (isDeferredApp) _buildAppUnitDropdown(currentTab.key!),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/app_size/app_size_screen.dart:211]: add some spacing here:

```dart
if (isDeferredApp) ...[
  _buildAppUnitDropdown(currentTab.key!),
  const SizedBox(width: defaultSpacing),
],
```

Comment (carolynqu) [reply packages/devtools_app/lib/src/screens/app_size/app_size_screen.dart:211]: Added spacing so that it only impacts the visuals when both drop downs are there.

File: packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart [hunk 6]
```diff
@@ -262,13 +262,15 @@ class AppSizeController {
     _diffCallGraphRoot.value = null;
     _oldDiffCallGraph = null;
     _newDiffCallGraph = null;
+    _isDeferredApp.value = false;
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/app_size/app_size_controller.dart:265]: Let's split this into a separate PR so we can add a test for it at the same time
